src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.LLVM.cs

```diff
@@ -2,38 +2,88 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Diagnostics;
-using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using System.Threading;
 
 namespace System
 {
     public partial class Exception
     {
-        internal static void DispatchExLLVM(object exception)
-        {
-            AppendExceptionStackFrameLLVM(exception, new StackTrace(1).ToString());
-            //RhpThrowEx(exception); can't as not handling the transition unmanaged->managed in the landing pads.
-        }
+        // Since we track stack traces as strings, we need an equivalent to the EdiSeparator marker.
+        // When set, we should treat "throws" as rethrows that do not reset the stack trace. This is
+        // an imperfect emulation of the real thing, as we don't "append" frames, but rather capture
+        // the full trace upfront, but it is better than nothing.
+        private bool _dispatchStateRestored;
 
-        private static void AppendExceptionStackFrameLLVM(object exceptionObj, string stackTraceString)
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal static void InitializeExceptionStackFrameLLVM(object exception, int flags)
         {
             // This method is called by the runtime's EH dispatch code and is not allowed to leak exceptions
             // back into the dispatcher.
             try
             {
-                Exception? ex = exceptionObj as Exception;
+                Exception? ex = exception as Exception;
                 if (ex == null)
                     Environment.FailFast("Exceptions must derive from the System.Exception class");
 
                 if (!RuntimeExceptionHelpers.SafeToPerformRichExceptionSupport)
                     return;
 
-                ex._stackTraceString = stackTraceString.Replace("__", ".").Replace("_", ".");
+                bool isFirstRethrowFrame = (flags & (int)RhEHFrameType.RH_EH_FIRST_RETHROW_FRAME) != 0;
+
+                // track count for metrics
+                if (!isFirstRethrowFrame)
+                    Interlocked.Increment(ref s_exceptionCount);
+
+                // When we're throwing an exception object, we reset its stacktrace with two exceptions:
+                // 1. Don't clear if we're rethrowing with `throw;`.
+                // 2. Don't clear if we're throwing through ExceptionDispatchInfo.
+                //    This is done through invoking RestoreDispatchState which sets "_dispatchStateRestored" followed by throwing normally using `throw ex;`.
+                if (isFirstRethrowFrame || ex._dispatchStateRestored)
+                    return;
+
+                // If out of memory, avoid any calls that may allocate.  Otherwise, they may fail
+                // with another OutOfMemoryException, which may lead to infinite recursion.
+                if (ex == PreallocatedOutOfMemoryException.Instance)
+                    return;
```

Comment:
Made the code look more like `AppendExceptionStackFrame` while I was here. We're now counting exceptions and guard against OOM recursion.